hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ReplicationManager.java

```diff
@@ -1314,4 +1774,257 @@ public boolean shouldRun() {
+        return this;
+      }
+
+      public Builder setDBTransactionBuffer(DBTransactionBuffer trxBuffer) {
+        transactionBuffer = trxBuffer;
+        return this;
+      }
+
+      public MoveScheduler build() throws IOException {
+        Preconditions.assertNotNull(moveTable, "moveTable is null");
+        Preconditions.assertNotNull(transactionBuffer,
+            "transactionBuffer is null");
+
+        final MoveScheduler impl =
+            new MoveSchedulerImpl(moveTable, transactionBuffer);
+
+        final SCMHAInvocationHandler invocationHandler
+            = new SCMHAInvocationHandler(MOVE, impl, ratisServer);
+
+        return (MoveScheduler) Proxy.newProxyInstance(
+            SCMHAInvocationHandler.class.getClassLoader(),
+            new Class<?>[]{MoveScheduler.class},
+            invocationHandler);
+      }
+    }
+  }
+
+  /**
+  * when scm become LeaderReady and out of safe mode, some actions
+  * should be taken. for now , it is only used for handle replicated
+  * infligtht move.
+  */
+  private void onLeaderReadyAndOutOfSafeMode() {
+    List<HddsProtos.ContainerID> needToRemove = new LinkedList<>();
+    moveScheduler.getInflightMove().forEach((k, v) -> {
+      Set<ContainerReplica> replicas;
+      ContainerInfo cif;
+      try {
+        replicas = containerManager.getContainerReplicas(k);
+        cif = containerManager.getContainer(k);
+      } catch (ContainerNotFoundException e) {
+        needToRemove.add(k.getProtobuf());
+        LOG.error("can not find container {} " +
+            "while processing replicated move", k);
+        return;
+      }
+      boolean isSrcExist = replicas.stream()
+          .anyMatch(r -> r.getDatanodeDetails().equals(v.getSrc()));
+      boolean isTgtExist = replicas.stream()
+          .anyMatch(r -> r.getDatanodeDetails().equals(v.getTgt()));
+
+      if(isSrcExist) {
+        if(isTgtExist) {
+          //the former scm leader may or may not send the deletion command
+          //before reelection.here, we just try to send the command again.
+          deleteSrcDnForMove(cif, replicas);
+        } else {
+          // resenting replication command is ok , no matter whether there is an
+          // on-going replication
+          sendReplicateCommand(cif, v.getTgt(),
+              Collections.singletonList(v.getSrc()));
+        }
+      } else {
+        // if container does not exist in src datanode, no matter it exists
+        // in target datanode, we can not take more actions to this option,
+        // so just remove it through ratis
+        needToRemove.add(k.getProtobuf());
+      }
+    });
+
+    needToRemove.forEach(moveScheduler::completeMove);
+  }
+
+  /**
+   * complete the CompletableFuture of the container in the given Map with
+   * a given MoveResult.
+   */
+  private void compleleteMoveFutureWithResult(ContainerID cid, MoveResult mr){
+    if(inflightMoveFuture.containsKey(cid)) {
+      inflightMoveFuture.get(cid).complete(mr);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **2024 and 2025**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **2026**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).